### ISSUE TITLE
Difference between sliders in lazer and stable

### ISSUE THREAD

**Issue by apollo-dw** [ppy/osu#18308]
### Type

Game behaviour

### Bug description

Slider generation is leading to two very severely different results between stable and lazer. Example sliders in the map: 00:21:142 (1,2,3,4) - 00:06:142 (1,2,3,4) - 

Really wacky slider node arrangement is likely the cause, 😕 

Map link: https://osu.ppy.sh/beatmapsets/1765052#osu/3612917
Alternative .osz download: https://cdn.discordapp.com/attachments/786626810086686740/975950870258802778/Sporty-O_-_Let_Me_Hit_It.osz

### Screenshots or videos

![screenshot103](https://user-images.githubusercontent.com/83023433/168717487-6e54ebc6-be90-4bc3-afea-401aef458b82.jpg)
![osu_2022-05-17_03-41-03](https://user-images.githubusercontent.com/83023433/168717506-698a6bec-74c2-4ec2-a628-01ce34d99792.jpg)

### Version

2022.515.0-lazer

### Logs

n/a

**Comment by peppy** [comment]
Is this present in any ranked beatmap? If not we may look to fix it in stable.

**Comment by smoogipoo** [comment]
Please let me investigate first before contributing to this discussion further. I know the cause and am doing a bunch of tests myself.

**Comment by smoogipoo** [comment]
This is caused by Catmull sliders. In stable, a Catmull slider with multiple "red" control points in it is treated as one segment. In lazer, these are treated as separate Catmull segments. This is basically a special case just for catmull sliders because all other types have multiple segments or don't support multiple segments at all (e.g. perfect-curve).

The differences can be summarised as:
| beatmap | stable | lazer |
| --- | --- | --- |
| [/b/22316](https://osu.ppy.sh/beatmapsets/3672#osu/22316) | ![stable_catmull](https://user-images.githubusercontent.com/1329837/168749142-5502b425-2007-429d-9bd5-96be307544d7.png) | ![lazer_catmull](https://user-images.githubusercontent.com/1329837/168749162-556a6d9c-9a72-445d-a415-0eeea97f98ec.png) |
| [/b/42223](https://osu.ppy.sh/beatmapsets/8499#osu/42223) | ![stable_catmull_2](https://user-images.githubusercontent.com/1329837/168749278-04921344-a7b8-4a47-bf14-44108c919600.png) | ![lazer_catmull_2](https://user-images.githubusercontent.com/1329837/168749297-137fe310-21df-4fb3-be19-ee6a9a8bc677.png) |

**Based on diffcalc alone**, here's a selection of the top SR differences in ranked beatmaps:
```
+------------+--------------+--------------+
| beatmap_id | diff_unified | diff_unified |
+------------+--------------+--------------+
|      42223 |      2.44393 |      2.42827 |
|      42223 |       2.3387 |      2.32657 |
|      26788 |      3.92047 |      3.93245 |
|      26788 |      5.22002 |      5.23179 |
|      42223 |      2.59259 |      2.58095 |
|      26788 |      3.25435 |      3.26595 |
|      42223 |      2.63576 |      2.62499 |
|      26788 |       3.8481 |      3.85839 |
|      42223 |      1.82955 |      1.81944 |
|      26788 |      3.19094 |      3.20093 |
|      26788 |      5.13611 |      5.14575 |
|      42223 |      2.46671 |      2.45768 |
|    2285424 |      13.4025 |      13.3935 |
|    2285424 |       12.982 |      12.9732 |
|    2285424 |      11.0776 |      11.0688 |
|    2285424 |      11.5443 |      11.5356 |
|      26788 |      2.68902 |      2.69753 |
|      42223 |      1.74935 |      1.74098 |
|      26788 |       3.2548 |      3.26314 |
|      42223 |      2.50824 |      2.49995 |
|      37571 |      1.49564 |      1.48742 |
|    2285424 |      9.81756 |      9.80981 |
|      42223 |      1.50159 |      1.49388 |
|    2285424 |      9.23291 |      9.22523 |
|      26788 |      4.36401 |      4.37167 |
|    2285424 |      9.81398 |      9.80637 |
|      26788 |        2.597 |      2.60443 |
|      42223 |      2.20168 |      2.19427 |
|    2285424 |      8.46535 |      8.45808 |
|    2285424 |      11.4627 |      11.4555 |
|      26788 |      3.15212 |      3.15932 |
|      37571 |       1.6008 |      1.59406 |
|      42223 |      1.96461 |      1.95802 |
|      37571 |      1.62765 |      1.62124 |
|      26788 |      4.24681 |      4.25317 |
|      42223 |      2.00272 |      1.99684 |
|      42223 |       1.4307 |      1.42489 |
|      36536 |      3.70021 |      3.69455 |
|      42223 |      1.86622 |       1.8607 |
|    2285424 |      9.95838 |      9.95289 |
|    2285424 |      8.56113 |      8.55572 |
|      42223 |      2.30392 |      2.29855 |
|      37571 |      1.36632 |      1.36099 |
|    2285424 |      8.15615 |      8.15085 |
|      36536 |      3.83543 |      3.83017 |
|    2285424 |      9.59581 |      9.59063 |
|      26788 |      2.67172 |      2.67687 |
|      26788 |      2.19499 |      2.20013 |
|      36536 |      3.87636 |      3.87122 |
|      42223 |      1.90304 |      1.89815 |
|      42223 |      2.34398 |      2.33918 |
|      26788 |       3.5857 |      3.59043 |
|    2285424 |      7.24205 |      7.23732 |
|    2285424 |      6.99884 |      6.99419 |
|    2285424 |      6.63958 |      6.63495 |
|    2285424 |      6.72987 |      6.72536 |
...
|    3358506 |      7.92029 |      7.92029 |
```
tl;dr: Seems to only be 2 beatmaps that have been ranked recently which experience differences. The older maps don't seem to be affected quite as much visually, however the differences on the recently-ranked ones are somewhat more impactful since they lead to partially-offscreen hitobjects or incorrect slider lengths:
| beatmap | stable | lazer |
| --- | --- | --- |
| [/b/2285424](https://osu.ppy.sh/beatmapsets/1093465#osu/2285424) | ![image](https://user-images.githubusercontent.com/1329837/168750794-f9fecc32-cd40-4c9a-8cf9-9f36f09c2b39.png) | ![image](https://user-images.githubusercontent.com/1329837/168750547-48e3a104-086b-474b-ab48-411cae2dd0ec.png) |
| [/b/3358506](https://osu.ppy.sh/beatmapsets/1561504#osu/3358506) | ![image](https://user-images.githubusercontent.com/1329837/168751786-3d23ca31-891e-4da4-a5a7-198cb6ea728c.png) | ![image](https://user-images.githubusercontent.com/1329837/168751721-919dd6a2-522d-45a2-b75d-b18618a2bffa.png) |

From the above two, only the first (`/b/2285424`) is impacted in a more meaningful way. The second only has 3 such sliders which break in the same (minor) fashion.

Here's a quick-and-dirty hack demonstrating the changes required to fix this in lazer:
```diff
diff --git a/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs b/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
index 2a7f2b037f..abdf1c1e8d 100644
--- a/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
+++ b/osu.Game/Rulesets/Objects/Legacy/ConvertHitObjectParser.cs
@@ -337,7 +337,7 @@ private IEnumerable<Memory<PathControlPoint>> convertPoints(ReadOnlyMemory<strin
             while (++endIndex < vertices.Length - endPointLength)
             {
                 // Keep incrementing while an implicit segment doesn't need to be started
-                if (vertices[endIndex].Position != vertices[endIndex - 1].Position)
+                if ((endIndex > 1 && type == PathType.Catmull) || vertices[endIndex].Position != vertices[endIndex - 1].Position)
                     continue;
 
                 // The last control point of each segment is not allowed to start a new implicit segment.
diff --git a/osu.Game/Rulesets/Objects/SliderPath.cs b/osu.Game/Rulesets/Objects/SliderPath.cs
index e0c62fe359..b0178f959f 100644
--- a/osu.Game/Rulesets/Objects/SliderPath.cs
+++ b/osu.Game/Rulesets/Objects/SliderPath.cs
@@ -221,7 +221,7 @@ private void calculatePath()
 
             for (int i = 0; i < ControlPoints.Count; i++)
             {
-                if (ControlPoints[i].Type == null && i < ControlPoints.Count - 1)
+                if ((ControlPoints[i].Type == null || ControlPoints[i].Type == PathType.Catmull) && i < ControlPoints.Count - 1)
                     continue;
 
                 // The current vertex ends the segment
```
It's possible that the change to `SliderPath` doesn't need to exist, however I'm not sure on this change in general as it would mean truly separate Catmull segments wouldn't be supported in lazer.

Curious on others' thoughts.

**Comment by peppy** [comment]
I'd be okay if it's only in the converter.

**Comment by smoogipoo** [comment]
@peppy That's only half of it. The question would be what to do for the editor, and what about beatmaps created via lazer.

Essentially, "red" control points wouldn't exist for catmull sliders.

**Comment by peppy** [comment]
There's already other breaking changes in the lazer editor regarding slider curves. I'd leave it as-is (ie. *not* apply the legacy behaviour to `vLazer` beatmaps). If we add support for mutli-type sliders back to stable this would also be supported in the process.
